synapse/cores/sortedram.py

```diff
@@ -0,0 +1,222 @@
+import collections
+
+import sortedcontainers
+
+import synapse.cores.common
+from synapse.compat import isint
+
+
+class SortedRamCortex(synapse.cores.common.Cortex):
+    '''
+    SortedRamCortex is a RAM-backed Cortex that maintain int prop values
+     in sorted order to facilitate range queries.
+
+    Insertion performance appears to be a constant factor of 10% slower than
+     synapse.cores.ram.Cortex
+    Query by range performance is strictly better.
+    Query by range runtime is O(log n) vs O(n) for the ram:// cortex.
+    '''
+
+    def _initCortex(self):
+
+        def getValu(row):
+            return row[2]
+
+        def makePropSet():
+            return sortedcontainers.SortedSet(key=getValu)
+
+        self.rowsbyid = collections.defaultdict(set)
+        # rows by props with type(value) == string
+        self.srowsbyprop = collections.defaultdict(set)
+        # rows by props with type(value) == int
+        # split because sortedcontainers.SortedSet does not support keys with mixed types
+        self.irowsbyprop = collections.defaultdict(makePropSet)
```

Comment:
I could see moving this to a function ( including the inner functions above ) since a "rowsorter()" may be something reusable or generalizable.